File: week-24/protoseo/동전.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine());
            int[] dp = new int[10001];
            int[] coins = new int[n];

            StringTokenizer stk = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) {
                coins[i] = Integer.parseInt(stk.nextToken());
            }

            dp[0] = 1;
            int m = Integer.parseInt(br.readLine());
            for (int coin : coins) {
                for (int j = coin; j <= m; j++) {
                    dp[j] += dp[j - coin];
                }
            }
            sb.append(dp[m]).append('\n');
        }
        System.out.println(sb);
    }
}

File: week-24/pushedrumex/동전.py
T = int(input())

for _ in range(T):
    N = int(input())
    coins = list(map(int, input().split()))
    M = int(input())

    dp = [0] * (M+1)
    dp[0] = 1

    # coin 종류 하나씩 사용
    for coin in coins:
        # coin 을 포함하여 만들수 있는 금액 + 1
        for m in range(coin, M+1):
            dp[m] += dp[m-coin]

    print(dp[M])